Mathlib/Topology/Algebra/Valued/ValExtension.lean

```diff
@@ -0,0 +1,245 @@
+/-
+Copyright (c) 2024 Patrick Massot. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jiedong Jiang, Bichang Lei
+-/
+import Mathlib.Topology.Algebra.Valued.ValuedField
+
+/-!
+# Extension of Valuation
+
+In this file, we define the typeclass for valuation extensions and prove basic facts about
+extension of valuations. Let `A` be a `R` algebra, both `R` and `A` are equipped with valued
+instance. Here, extension of valuation means that the pullback of valuation on `A` is equivalent
+to the valuation on `R`. We only require equivalence, not equality of valuations here.
+
+Note that we do not require the ring map from `R` to `A` to be injective. It holds automatically
+when `R` is a division ring and `A` is nontrivial.
+
+## Main Definition
+
+* `IsValExtension R A` : The valuation on `A` is an extension of the valuation on `R`.
+
+## References
+
+* [Bourbaki, Nicolas. *Commutative algebra*] Chapter VI §3, Valuations.
+* <https://en.wikipedia.org/wiki/Valuation_(algebra)#Extension_of_valuations>
+
+## Tags
+Valued, Valuation, Extension of Valuation
+
+-/
+open Valued Valuation
+
+namespace Valued
+
+variable {R S : Type*} {ΓR ΓS : outParam Type*} [Ring R] [Ring S]
+    [LinearOrderedCommGroupWithZero ΓR] [LinearOrderedCommGroupWithZero ΓS]
+    [vR : Valued R ΓR] [vS : Valued S ΓS]
+
+variable {F : Type*} [FunLike F R S] [RingHomClass F R S] {f : F}
+    (hf : vR.v.IsEquiv <| vS.v.comap f)
+
+theorem val_map_le_iff (x y : R) : v (f x) ≤ v (f y) ↔ v x ≤ v y :=
+  (hf x y).symm
+
+theorem val_map_lt_iff (x y : R) : v (f x) < v (f y) ↔ v x < v y := by
+  convert (val_map_le_iff hf y x).not <;>
+  push_neg <;> rfl
+
+theorem val_map_eq_iff (x y : R) : v (f x) = v (f y) ↔ v x = v y := by
+  calc
+    _ ↔ v (f x) ≤ v (f y) ∧ v (f y) ≤ v (f x) := le_antisymm_iff
+    _ ↔ v x ≤ v y ∧ v y ≤ v x := by
+      apply and_congr <;>
+      exact val_map_le_iff hf _ _
+    _ ↔ _ := le_antisymm_iff.symm
+
+theorem val_map_le_one_iff (x : R) : v (f x) ≤ 1 ↔ v x ≤ 1 := by
+  convert val_map_le_iff hf x 1 <;>
+  simp only [_root_.map_one]
+
+theorem val_map_lt_one_iff (x : R) : v (f x) < 1 ↔ v x < 1 := by
+  convert val_map_lt_iff hf x 1 <;>
+  simp only [_root_.map_one]
+
+theorem val_map_eq_one_iff (x : R) : v (f x) = 1 ↔ v x = 1 := by
+  convert val_map_eq_iff hf x 1 <;>
+  simp only [_root_.map_one]
+
+end Valued
+
+/--
+The class `IsValExtension R A` states that the valuation of `A` is an extension of the valuation
+on `R`. More precisely, the valuation on `R` is equivlent to the comap of the valuation on `A`.
+-/
+class IsValExtension (R A : Type*) {ΓR ΓA : outParam Type*} [CommRing R] [Ring A]
+  [LinearOrderedCommGroupWithZero ΓR] [LinearOrderedCommGroupWithZero ΓA]
+  [Algebra R A] [vR : Valued R ΓR] [vA : Valued A ΓA] : Prop where
+  /-- The valuation on `R` is equivlent to the comap of the valuation on `A` -/
+  val_isEquiv_comap : vR.v.IsEquiv <| vA.v.comap (algebraMap R A)
```

Comment:
I think this (and many results below) does not actually need the topological structure and should use `Valuation` instead of `Valued` and go in `RingTheory/Valuation`.

Reply:
I agree that these results do not actually need the topological structure. The idea of choice here is that, for convenience, I want to make this into a typeclass so that Lean can infer it automatically. (Since almost all field extensions that people care about in the local case are extensions of valuation.) `Valued` is a class, but `Valuation` is not a class. What do you suggest?

Reply:
What about
```
class IsValExtension {R A ΓR ΓA : Type*} [CommRing R] [Ring A]
  [LinearOrderedCommGroupWithZero ΓR] [LinearOrderedCommGroupWithZero ΓA]
  [Algebra R A] (vR : Valuation R ΓR) (vA : Valuation A ΓA) : Prop where
  /-- The valuation on `R` is equivlent to the comap of the valuation on `A` -/
  val_isEquiv_comap : vR.IsEquiv <| vA.comap (algebraMap R A)
```
So that `IsValExtension vR vA` still can be inferred automatically.